### ISSUE TITLE
UnionType is revealed instead of actual union

### ISSUE THREAD

**Issue by AlexanderPodorov** [microsoft/pyright#7272]
**Describe the bug**
See example code.

```python
from pydantic import TypeAdapter

# type of "ta" is "UnionType", should be "int | str"
ta = TypeAdapter(int | str).validate_python(4)

# type of "ta" is int
ta = TypeAdapter(int).validate_python(4)

```

`pydantic==2.6.1` is used

**VS Code extension or command-line**
Pylance v2024.2.101 (pre-release), shipped with pyright 1.1.350

**Comment by erictraut** [comment]
There is a bug here, but it's not the bug you're reporting.

The first example should generate an error because `int | str` when used in a value expression is an instance of `UnionType`, and this is not compatible with `type[T]`. 

I've been working to increase pyright's conformance with the Python typing spec. This was an area where pyright was previously too lax. By contrast, mypy correctly generates an error for this first case in your example.

There has been some discussion of introducing a new `TypeForm[T]` annotation to the type system that accommodates this use case, but it hasn't yet made it to the draft PEP stage.

**Comment by erictraut** [comment]
Digging a little deeper, pyright is doing the right thing here. That's because the constructor for pydantic's `TypeAdapter` class is overloaded. The second overload accepts an objects of arbitrary type.

```python
    @overload
    def __init__(  # pyright: ignore[reportOverlappingOverload]
        self,
        type: T,
        *,
        config: ConfigDict | None = ...,
        _parent_depth: int = ...,
        module: str | None = ...,
    ) -> None:
        ...
```

This second overload is compatible with an object of type `UnionType`, so pyright is correct to not generate an error here.

I'm closing the issue because I think pyright is doing the right thing.
